internal/armada/queryapi/query_api.go

```diff
@@ -2,51 +2,208 @@ package queryapi
+				LastTransitionTs: DbTimeToGoTime(row.LastTransitionTime),
+				LatestRunId:      NilStringToString(row.LatestRunID),
+				JobSpec:          jobSpec,
+			}
+			if req.GetExpandJobRun() && row.LatestRunID != nil {
+				jobsWithRuns = append(jobsWithRuns, row.JobID)
+			}
+		}
+		// Fetch the Job run details in a separate query.
+		// We do this because each job can have many runs and so we don;t want to duplicate the job data for each run
+		if len(jobsWithRuns) > 0 {
+			runResultRows, err := queries.GetJobRunsByJobIds(ctx, jobsWithRuns)
+			if err != nil {
+				return err
+			}
+			runsByJob := make(map[string][]*api.JobRunDetails, len(resultRows))
+			for _, row := range runResultRows {
+				jobRuns, ok := runsByJob[row.JobID]
+				if !ok {
+					jobRuns = []*api.JobRunDetails{}
+				}
+				jobRuns = append(jobRuns, parseJobDetails(row))
+				runsByJob[row.JobID] = jobRuns
+			}
+
+			for jobId, jobDetails := range detailsById {
+				runs, ok := runsByJob[jobId]
+				if ok {
+					jobDetails.JobRuns = runs
+				}
+			}
+		}
+		return nil
+	})
+
+	return &api.JobDetailsResponse{
+		JobDetails: detailsById,
+	}, err
+}
+
+func (q *QueryApi) GetJobRunDetails(ctx context.Context, req *api.JobRunDetailsRequest) (*api.JobRunDetailsResponse, error) {
+	if len(req.RunIds) > q.maxQueryItems {
+		return nil, fmt.Errorf("request contained more than %d RunIds", q.maxQueryItems)
+	}
+
+	queries := database.New(q.db)
+	resultRows, err := queries.GetJobRunsByRunIds(ctx, req.RunIds)
+	if err != nil {
+		return nil, err
+	}
+	detailsById := make(map[string]*api.JobRunDetails, len(resultRows))
+	for _, row := range resultRows {
+		detailsById[row.RunID] = parseJobDetails(row)
+	}
+	return &api.JobRunDetailsResponse{
+		JobRunDetails: detailsById,
+	}, nil
 }
 
 func (q *QueryApi) GetJobStatus(ctx context.Context, req *api.JobStatusRequest) (*api.JobStatusResponse, error) {
+	if len(req.JobIds) > q.maxQueryItems {
+		return nil, fmt.Errorf("request contained more than %d jobIds", q.maxQueryItems)
+	}
+
 	queries := database.New(q.db)
-	queryResult, err := queries.GetJobState(ctx, req.JobId)
+	queryResult, err := queries.GetJobStates(ctx, req.JobIds)
 	if err != nil {
 		return nil, err
 	}
-	status := int16(-1)
-	if len(queryResult) > 0 {
-		status = queryResult[0]
+	dbStatusById := make(map[string]int16, len(queryResult))
+	for _, dbRow := range queryResult {
+		dbStatusById[dbRow.JobID] = dbRow.State
 	}
-	apiStatus, ok := JobStateMap[status]
-	if !ok {
-		apiStatus = api.JobStatus_UNKNOWN
+
+	apiStatusById := make(map[string]api.JobState, len(queryResult))
+	for _, jobId := range req.JobIds {
+		dbStatus, ok := dbStatusById[jobId]
+		if ok {
+			apiStatus, ok := JobStateMap[dbStatus]
+			if !ok {
+				apiStatus = api.JobState_UNKNOWN // We know about this job but we can't map its state
+			}
+			apiStatusById[jobId] = apiStatus
+		} else {
+			apiStatusById[jobId] = api.JobState_UNKNOWN // We don't know about this job
```

Comment:
Will we ever need a way to differentiate between these 2 cases? I guess in theory the first case should never happen

Reply:
The first case should really be an error and I might move it to be one once I'm more confident.  For now I'll just log a warning.